nevergrad/optimization/optimizerlib.py

```diff
@@ -1534,6 +1540,220 @@ def __init__(
+
+        self.trials = Trials()
+        self.domain = Domain(fn=None, expr=self.space, pass_expr_memo_ctrl=False)
+        self.tpe_args = {
+            "prior_weight": prior_weight,
+            "n_startup_jobs": n_startup_jobs,
+            "n_EI_candidates": n_EI_candidates,
+            "gamma": gamma,
+            "verbose": verbose
+        }
+
+    def _get_search_space(self, param_name, param):
+        if isinstance(param, p.Instrumentation):
+            space = {}
+            space["args"] = {
+                    str(idx_param): self._get_search_space(str(idx_param), param[0][idx_param]) # type: ignore
+                    for idx_param in range(len(param[0].value))
+            }
+            space["kwargs"] = {
+                    param_name: self._get_search_space(param_name, param[1][param_name]) # type: ignore
+                    for param_name in param[1].value.keys()
+            }
+            return space
+        elif isinstance(param, (p.Log, p.Scalar)):
+            if (param.bounds[0][0] is None) or (param.bounds[1][0] is None):
+                raise ValueError(f"Scalar {param_name} not bounded.")
+            elif isinstance(param, p.Log):
+                return hp.loguniform(label=param_name,
+                                     low=np.log(param.bounds[0][0]),
+                                     high=np.log(param.bounds[1][0]))
+            elif isinstance(param, p.Scalar):
+                if param.integer:
+                    return hp.randint(label=param_name,
+                                      low=int(param.bounds[0][0]),
+                                      high=int(param.bounds[1][0]))
+                else:
+                    return hp.uniform(label=param_name,
+                                      low=param.bounds[0][0],
+                                      high=param.bounds[1][0])
+        elif isinstance(param, p.Choice):
+            list_types = [type(param.choices[i]) for i in range(len(param.choices))
+                          if not isinstance(param.choices[i], (p.Instrumentation, p.Constant))]
+
+            if len(list_types) != len(set(list_types)): raise NotImplementedError
+            return hp.choice(param_name,
+                             [self._get_search_space(param_name + "__" + str(i), param.choices[i])
+                              for i in range(len(param.choices))])
+        elif isinstance(param, p.Constant):
+            return param.value
+
+        # Hyperopt do not support array
+        raise NotImplementedError
+
+    def _dict_to_parametrization(self, x):
+        if isinstance(x, dict) and ("args" in x) and ("kwargs" in x):
+            x["args"] = tuple([self._dict_to_parametrization(x["args"][str(i)]) for i in range(len(x["args"]))])
+            x["kwargs"] = {k: self._dict_to_parametrization(v) for k, v in x["kwargs"].items()}
+            return (x["args"], x["kwargs"])
+        return x
+
+    def _parametrization_to_dict(self, x, **kwargs):
```

Comment:
since self is not really used in `_dict_to_parametrization` and `_parametrization_to_dict` it would be better to move them out of the class (and keep them in the safe hyperopt-specific file as mentioned aboove), and test them independently of the rest